### ISSUE TITLE
Use scikit-optimize for tuning search space for model selection

### ISSUE THREAD

**Issue by mnarayan** [skggm/skggm#95]
Not being beholding to specifying a sub-optimal grid/family of models for tuning the initial estimator will be extremely useful. 

Use drop-in replacement for `GridSearchCV` from [scikit-optimize](https://scikit-optimize.github.io/).

**Comment by mnarayan** [comment]
This is how one would invoke `BayesSearchCV` using QUIC's default mode. 
```python
    model = BayesSearchCV(QuicGraphLasso(init_method='corrcoef', score_metric=metric),
                         cv=num_folds,
                         n_iter=16,
                         refit=True
                         ) 
    model.add_spaces('space_1', {
                        'lam': Real(1e-02,1e+1, prior='log-uniform')
                        }
                    )
    model.fit(X)
```
Here is an [example](https://gist.github.com/mnarayan/1c4287e3b8f8105c218c2150a8e5e7b4#file-skggm_optimal_selection_example-py) of using `BayesSearchCV` 

Results from running `run skggm_optimal_selection_example.py` output
```
QuicGraphLasso + GridSearchCV with:
   metric: kl
   len(cv_lams): 100
   cv-lam: 0.148496826225
   lam_scale_: 1.0
   lam_: 0.148496826225
QuicGraphLasso + skopt.BayesSearchCV with:
   metric: kl
   Final score: -0.224238139648
   cv-lam: 0.139651618375
   lam_scale_: 1.0
   lam_: 0.139651618375
```

**Comment by jasonlaska** [comment]
Based on your example here, this feels more like something that should go in `/examples` and not something that should be part of the main package.  My reasoning is:

- we should shy away from providing light wrappers over other packages (or just combinations of packages) since then we have increased maintenance when they change their interface, introduce incompatibilities, etc.  It's fine to provide an extensive set of examples using these, but otherwise, where would we draw the line in terms of combinations of helper interfaces.

- if we needed to specially modify the QuicGraphLasso to take advantage of this package (such as the more optimized QuicGraphLassoCV), that would be different, but based on your example and the way BayesSearchCV works, it doesn't seem that we need to.

- I do think that a rich set of examples on how to use the tools and best practices makes sense, but it need not be on the mainline package.

**Comment by mnarayan** [comment]
@jasonlaska The example I made was just a brute force example to show usage of `BayesSearchCV` which is a wrapper around skopt's `Optimizer` class.  
The real changes I think need to happen as you pointed out here

>- if we needed to specially modify the QuicGraphLasso to take advantage of this package (such as the more optimized QuicGraphLassoCV), that would be different, but based on your example and the way BayesSearchCV works, it doesn't seem that we need to.

skopt's Optimizer lets you go about adaptively picking a grid of points to evaluate.  For example, this would not be possible to achieve to accomplish for EBIC by invoking `BayesSearchCV`.  We would need to model our own QuicGraphLassoCV and QuicGraphLassoEBIC similar to BayesSearchCV.  Presumably, we could analogously take advantage of path mode as well, just like you enabled with QuicGraphLassoCV.

The moment we have more than one path to optimize over multiple penalties or multiple parameters,  this starts to become a lot more handy as well.
